randomizer/residual_push.py

```diff
@@ -0,0 +1,246 @@
+    """Superclass for all Fetch environments.
+    """
+
+    def __init__(
+            self, model_path, n_substeps, gripper_extra_height, block_gripper,
+            has_object, target_in_the_air, target_offset, obj_range, target_range,
+            distance_threshold, initial_qpos, reward_type, **kwargs,
+    ):
+        """Initializes a new Fetch environment.
+
+        Args:
+            model_path (string): path to the environments XML file
+            n_substeps (int): number of substeps the simulation runs on every call to step
+            gripper_extra_height (float): additional height above the table when positioning the gripper
+            block_gripper (boolean): whether or not the gripper is blocked (i.e. not movable) or not
+            has_object (boolean): whether or not the environment has an object
+            target_in_the_air (boolean): whether or not the target should be in the air above the table or on the table surface
+            target_offset (float or array with 3 elements): offset of the target
+            obj_range (float): range of a uniform distribution for sampling initial object positions
+            target_range (float): range of a uniform distribution for sampling a target
+            distance_threshold (float): the threshold after which a goal is considered achieved
+            initial_qpos (dict): a dictionary of joint names and values that define the initial configuration
+            reward_type ('sparse' or 'dense'): the reward type, i.e. sparse or dense
+        """
+
+        self.gripper_extra_height = gripper_extra_height
+        self.block_gripper = block_gripper
+        self.has_object = has_object
+        self.target_in_the_air = target_in_the_air
+        self.target_offset = target_offset
+        self.obj_range = obj_range
+        self.target_range = target_range
+        self.distance_threshold = distance_threshold
+        self.reward_type = reward_type
+
+        super(RandomizedResidualPushEnv, self).__init__(
+            model_path=model_path, n_substeps=n_substeps, n_actions=4,
+            initial_qpos=initial_qpos)
+
+        # randomization
+        self.xml_dir = os.path.join(os.path.dirname(os.path.realpath(__file__)), "assets_residual")
+        self.reference_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), "assets_residual",
+                                           kwargs.get('xml_name'))
+        self.reference_xml = et.parse(self.reference_path)
+        self.config_file = kwargs.get('config')
+        self.dimensions = []
+        self.dimension_map = []
+        self.suffixes = []
+        self._locate_randomize_parameters()
+
+
+    # Randomization methods
+    # ----------------------------
+    def _locate_randomize_parameters(self):
+        self.root = self.reference_xml.getroot()
+        self.object_joints = self.root.findall(".//body[@name='object0']/joint")
+
+    def _randomize_damping(self):
```

Comment:
So damping is just an example, but what we actually want to randomize is friction, which you will set via `_randomize_friction()` and then set it like:

```python

for i in range(len(self.fetch_env.env.sim.model.geom_friction)):
   self.fetch_env.env.sim.model.geom_friction[i] = [18e-2 * MULTIPLIER | CURRENT VALUE, 5.e-3, 1e-4]
```